library/alloc/src/collections/btree/node.rs

```diff
@@ -342,28 +342,36 @@ impl<BorrowType, K, V, Type> NodeRef<BorrowType, K, V, Type> {
     /// # Safety
     /// The node has more than `idx` initialized elements.
     pub unsafe fn key_at(&self, idx: usize) -> &K {
```

Comment:
This function is not even used any more in this file... is it even worth keeping (as opposed to expecting the caller to handle `MaybeUninit`?

Reply:
Well it's there for `search_linear`?

Reply:
At first I found it weird to have a function returning a reference to something that you can only use assuming it's initialized. But it's there because I didn't know how else to copy/write a slice into uninitialized elements. And it's somewhat nice to have functions that only do one thing.

Reply:
> Well it's there for search_linear? in search.rs

If that's the only user, it can just to `assume_init_ref` itself though?

> At first I found it weird to have a function returning a reference to something that you can only use assuming it's initialized. But it's there because I didn't know how else to copy/write a slice into uninitialized elements. And it's somewhat nice to have functions that only do one thing.

Well, I do not understand this code enough to suggest other APIs here, or to tell how many of these callers call this function on uninit elements.

Reply:
Actually, it doesn't seem useful to change `key_at` after all. To copy a slice from one node to another, miri-track-raw-pointers wants the source pointer to point something it can read entirely, not just the first element. Which is sensible and easy to do, but it leaves `key_at` only useful to read a key.

The next problem, to which I don't see a way out at the moment, is `as_internal_mut()`:

```
302 |         unsafe { &mut *(self.node.as_ptr() as *mut InternalNode<K, V>) }
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ trying to reborrow for Unique at alloc22839+0x140, but parent tag <58889> does not have an appropriate item in the borrow stack
    |
```

We disguise a `*mut InternalNode` as a `*mut LeafNode` to squeeze it into `BoxedNode`, and there's no way around that. Now how do we unpack such a party pooper and convince Miri that this is an InternalNode?

-Zmiri-track-pointer-tag tracks the tag to the creator, `BoxedNode::from_internal`:
```
BoxedNode { ptr: Unique::from(&mut Box::leak(node).data) }
```

It does not help to make the reference raw:
```
`BoxedNode { ptr: unsafe { Unique::new_unchecked(&raw mut Box::leak(node).data) } }
```

Reply:
> Actually, it doesn't seem useful to change key_at after all. To copy a slice from one node to another, miri-track-raw-pointers wants the source pointer to point something it can read entirely, not just the first element. Which is sensible and easy to do, but it leaves key_at only useful to read a key.

Makes sense.

> We disguise a *mut InternalNode as a *mut LeafNode to squeeze it into BoxedNode, and there's no way around that. Now how do we unpack such a party pooper and convince Miri that this is an InternalNode?

Everything along the line needs to be raw pointers.

> -Zmiri-track-pointer-tag tracks the tag to the creator, BoxedNode::from_internal:

Does it make any difference if you cherry-pick https://github.com/rust-lang/rust/pull/78446? I think it should not, but the error points at the same place...

> -Zmiri-track-pointer-tag tracks the tag to the creator, BoxedNode::from_internal:

Could you show the full output with pointer tracking, i.e. all the tracking messages and the final error?

Reply:
> Everything along the line needs to be raw pointers.

It seems to work when casting in both directions. I thought I'd made things more clear by avoiding that cast, but clarity steered Miri up the wrong alley.

> Does it make any difference if you cherry-pick #78446?

I'm doing this on Miri's current rust-version, before the allocator upheaval, with your #78476 rebased. I don't understand how you seem to let Miri digest the latest std. Perhaps you just glance over the code in your head and project those issues, I wouldn't be surprised.

Reply:
> I'm doing this on Miri's current rust-version, before the allocator upheaval, with your #78476 rebased.

Oh right, then that is no concern.

> I don't understand how you seem to let Miri digest the latest std.

At least for some of these investigations I ended up building Miri against a locally built rustc toolchain (as I was patching the core Miri engine to help debugging... it didn't end up helping though^^).

Reply:
> It seems to work when casting in both directions. I thought I'd made things more clear by avoiding that cast, but clarity steered Miri up the wrong alley.

What is the alternative to casting?

Reply:
> What is the alternative to casting?

Like the code above: access the first member `.data` of this `repr(C)` struct.

Reply:
Ah I see. The problem wasn't with the member access, the problem was with the `&mut`. That violates the "only use raw pointers" rule it creates a reference.

Using `core::ptr::raw_mut!(something.data)` instead should also work.

Reply:
But that looks suspiciously like the `&raw mut something.data` I tried above in "It does not help to make the reference raw". I just tried it again, and get the same complaint. The pointer tracking says it all:
```
126 |         BoxedNode { ptr: unsafe { Unique::new_unchecked(ptr::raw_mut!(Box::leak(node).data)) } }
    |                                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ created tag 58937
    |
```
It's remembering the data. Or actually, it's just #74679 again?

Reply:
Nope. Given 
```rust
    fn from_internal(node: Box<InternalNode<K, V>>) -> Self {
        let whole = ptr::raw_mut!(*Box::leak(node));
        let part = unsafe { ptr::raw_mut!((*whole).data) };
        BoxedNode { ptr: unsafe { Unique::new_unchecked(part) } }
    }
```
the tag of the internal pointer still refers to:
```
    |
127 |         let part = unsafe { ptr::raw_mut!((*whole).data) };
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ created tag 58938
```

Reply:
That is strange, I will have to investigate.

Reply:
It does remind me of [a much simpler example](https://users.rust-lang.org/t/why-does-raw-field-pointer-access-trigger-miri-to-pop-a-unique/50449) I've been trying to unravel.

Reply:
Same behaviour with this example:
```rust
#![feature(ptr_internals, raw_ref_macros)]
use std::ptr::{self, NonNull, Unique};

struct Part {
    _lame: i32,
}

#[repr(C)]
struct Whole {
    part: Part,
    extra: i32,
}

fn main() {
    let it = Box::new(Whole { part: Part { _lame: 0 }, extra: 42 });
    let whole = ptr::raw_mut!(*Box::leak(it));
    let part = unsafe { ptr::raw_mut!((*whole).part) };
    let unique = unsafe { Unique::new_unchecked(part) };
    let raw = NonNull::from(unique);
    let typed = unsafe { &mut *(raw.as_ptr() as *mut Whole) };
    assert!(typed.extra == 42);
}
```
It also disappears by removing the _lame field.

Reply:
Yeah this is a Miri / Stacked Borrows problem, good catch! See https://github.com/rust-lang/miri/issues/1608.